hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/FromElementType.java

```diff
@@ -685,7 +687,27 @@ public Type toType(String propertyName) throws QueryException {
 	public String getIdentifierPropertyName() {
 		if ( getEntityPersister() != null && getEntityPersister().getEntityMetamodel() != null
 				&& getEntityPersister().getEntityMetamodel().hasNonIdentifierPropertyNamedId() ) {
-			return getEntityPersister().getIdentifierPropertyName();
+			if ( getEntityPersister().hasIdentifierProperty() ) {
+				return getEntityPersister().getIdentifierPropertyName();
+			}
+			else {
+				final IdentifierProperty identifierProperty =
+						getEntityPersister().getEntityMetamodel().getIdentifierProperty();
+				if ( identifierProperty.hasIdentifierMapper() && !identifierProperty.isEmbedded() ) {
+					return PropertyPath.IDENTIFIER_MAPPER_PROPERTY;
+				}
+				else {
+					throw new UnsupportedOperationException(
+							String.format(
+									"[%s] has a composite key and a non-ID property named 'id'; " +
+											"Hibernate only supports this use case when " +
+											"the composite ID is mapped using a primary key class.",
+									getEntityPersister().getEntityName()
+							)
+
+					);
+				}
+			}
```

Comment:
I'm surprised this logic is only valid for this use case? All the other `EntityPersister#getIdentifierPropertyName()` calls in the code base are safe?